File: src/045/p045.py
import math
from src.utils.py.series import pentagonal_numbers, triangle_numbers


def is_pentagonal_number(number: int) -> bool:
    root = math.sqrt(24 * number + 1)
    return root % 1 == 0 and (int(root) + 1) % 6 == 0


def is_hexagonal_number(number: int) -> bool:
    root = math.sqrt(8 * number + 1)
    return root % 1 == 0 and (int(root) + 1) % 4 == 0


def solve1() -> int:
    for num in triangle_numbers():
        if num > 40755 and is_pentagonal_number(num) and is_hexagonal_number(num):
            return num
    raise RuntimeError("No solution found")


def solve() -> int:
    for num in pentagonal_numbers():
        if num > 40755 and is_hexagonal_number(num):
            return num
    raise RuntimeError("No solution found")

File: src/045/p045.go
package problem045

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math"
	"github.com/sunilbpandey/project-euler/src/utils/go/math/intutils"
)

func isHexagonal(n int) bool {
	sqrt := intutils.Sqrt(8*n + 1)
	return sqrt*sqrt == 8*n+1 && sqrt%4 == 3
}

func Solve() string {
	p := 0
	for p = range math.PentagonalNumbers() {
		if p > 40755 && isHexagonal(p) {
			break
		}
	}
	return strconv.Itoa(p)
}
